xdesign/propagation.py

```diff
@@ -0,0 +1,234 @@
+    # H2 = fftshift(fftn(h))
+    # print(2)
+    # plt.figure()
+    # plt.imshow(np.abs(H))
+    # plt.figure()
+    # plt.imshow(np.abs(H2))
+    # plt.show()
+
+    # time.sleep(999)
+
+    wavefront = ifftn(ifftshift(fftshift(fftn(wavefront)) * H))
+    # H = np.exp(-1j * 2 * np.pi * delta_nm / lmda_nm * np.sqrt(1. - lmda_nm ** 2 * u ** 2))
+    # wavefront = np.fft.ifftn(np.fft.fftn(wavefront) * np.fft.fftshift(H))
+    return wavefront
+
+
+def far_propagate(grid, wavefront, dist_nm):
+    """Free space propagation using product Fourier algorithm. Suitable for far field propagation.
+    """
+    assert isinstance(grid, Grid3d)
+    lmbda_nm = grid.lmbda_nm
+    u_max = 1. / (2. * grid.voxel_x)
+    v_max = 1. / (2. * grid.voxel_y)
+    u, v = gen_mesh([v_max, u_max], grid.grid_delta.shape[1:3])
+    x = grid.xx[0, :, :] * grid.voxel_x
+    y = grid.yy[0, :, :] * grid.voxel_y
+    x_max = grid.size[0] * grid.voxel_x
+    y_max = grid.size[1] * grid.voxel_y
+    # h = np.exp(-1j * 2 * np.pi / lmbda_nm * np.sqrt(dist_nm ** 2 + x ** 2 + y ** 2))
+    # wavefront = fftshift(fft2(wavefront * h))
+    # wavefront = wavefront * \
+    #             np.exp(-1j * dist_nm / lmbda_nm * np.sqrt(4 * np.pi ** 2 + lmbda_nm ** 2 * (u ** 2 + v ** 2))) * \
+    #             4 * np.pi / (x_max ** 2 + y_max ** 2) / (lmbda_nm * dist_nm) * 1j
+
+    h = np.exp(-1j * np.pi * (x ** 2 + y ** 2) / (lmbda_nm * dist_nm))
+    wavefront = fftshift(fft2(wavefront * h))
+    wavefront = wavefront * np.exp(-1j * np.pi * lmbda_nm * dist_nm * (u ** 2 + v ** 2))
+    wavefront = wavefront * 1j / (lmbda_nm * dist_nm)
+
+    print(u)
+    print(lmbda_nm, dist_nm)
+    print(u * lmbda_nm * dist_nm)
+
+    # y0 = grid.yy[0, :, :]
+    # x0 = grid.xx[0, :, :]
+    # y = y0 * (lmda * z) * (grid.size[1] * grid.voxel_y) ** 2
+    # x = x0 * (lmda * z) * (grid.size[0] * grid.voxel_x) ** 2
+    # wavefront = fftshift(fftn(wavefront * np.exp(-1j * 2 * np.pi / lmda * np.sqrt(z ** 2 + x0 ** 2 + y0 ** 2))))
+    # wavefront = wavefront * np.exp(-1j * 2 * np.pi / lmda * np.sqrt(z ** 2 + x ** 2 + y ** 2))
+    return wavefront
+
+
+def _far_propagate_2(grid, wavefront, lmd, z_um):
+    """Free space propagation using product Fourier algorithm.
+    """
+    assert isinstance(grid, Grid3d)
+
+    N = grid.size[1]
+    M = grid.size[2]
+    D = N * grid.voxel_y
+    H = M * grid.voxel_x
+    f1 = wavefront
+
+    V = N/D
+    U = M/H
+    d = np.arange(-(N-1)/2,(N-1)/2+1,1)*D/N
+    h = np.arange(-(M-1)/2,(M-1)/2+1,1)*H/M
+    v = np.arange(-(N-1)/2,(N-1)/2+1,1)*V/N
+    u = np.arange(-(M-1)/2,(M-1)/2+1,1)*U/M
+
+    f2 = np.fft.fftshift(np.fft.fft2(f1*np.exp(-1j*2*np.pi/lmd*np.sqrt(z_um**2+d**2+h[:,np.newaxis]**2))))*np.exp(-1j*2*np.pi*z_um/lmd*np.sqrt(1.+lmd**2*(v**2+u[:,np.newaxis]**2)))/U/V/(lmd*z_um)*(-np.sqrt(1j))
+    d2,h2=v*lmd*z_um,u*lmd*z_um
+    return f2
+
+
+def plot_wavefront(wavefront, grid, save_folder='simulation', fname='exiting_wave'):
```

Comment:
Plot functions should all be in the `plot` module in order to keep import statements minimized.